crates/sparrow-scheduler/src/pipeline.rs

```diff
@@ -0,0 +1,128 @@
+/// a child index, and a partition index, whereas output partitions are only
+/// identified by a partition index.
+///
+/// This is not intended as an eventual replacement for the physical plan
+/// representation, but rather a generic interface that parts of the physical
+/// plan are converted to for execution.
+///
+/// # Eager vs Lazy Execution
+///
+/// Whether computation is eagerly done on push, or lazily done on pull, is
+/// intentionally left as an implementation detail of the [`Pipeline`]
+///
+/// This allows flexibility to support the following different patterns, and potentially more:
+///
+/// An eager, push-based pipeline, that processes a batch synchronously in [`Pipeline::push`]
+/// and immediately wakes the corresponding output partition.
+///
+/// A parallel, push-based pipeline, that enqueues the processing of a batch to
+/// the thread pool in [`Pipeline::push`], and wakes the corresponding output
+/// partition when the job completes. Order and non-order preserving variants
+/// are possible
+///
+/// A merge pipeline which combines data from one or more input partitions into one or
+/// more output partitions. [`Pipeline::push`] adds data to an input buffer, and wakes
+/// any output partitions that may now be able to make progress. This may be none if
+/// the operator is waiting on data from a different input partition.
+///
+/// An aggregation pipeline which combines data from one or more input partitions into
+/// a single output partition. [`Pipeline::push`] would eagerly update the computed
+/// aggregates, and the final [`Pipeline::close`] trigger flushing these to the output.
+/// It would also be possible to flush once the partial aggregates reach a certain size.
+///
+/// A partition-aware aggregation pipeline, which functions similarly to the above, but
+/// computes aggregations per input partition, before combining these prior to flush.
+///
+/// An async input pipeline, which has no inputs, and wakes the output partition
+/// whenever new data is available.
+///
+/// A JIT compiled sequence of synchronous operators, that perform multiple operations
+/// from the physical plan as a single [`Pipeline`]. Parallelized implementations
+/// are also possible.
+///
+pub trait Pipeline: Send + Sync + std::fmt::Debug {
+    /// Initialize the pipeline by providing the tasks it will use for scheduling.
+    ///
+    /// The number of tasks provided indicate the number of partitions this pipeline
+    /// should execute with.
+    ///
+    /// This is unfortunately separate to allow cyclic initialization. Specifically,
+    /// `Arc::new_cyclic` expects a function that doesn't take errors, while creating
+    /// many pipelines *does* produce errors. To address this, we first create the
+    /// part of the pipeline that doesn't need to reference it's own tasks, and then
+    /// we initialize it as part of `Arc::new_cyclic`.
+    fn initialize(&mut self, tasks: Partitioned<TaskRef>);
```

Comment:
Do you have any future concerns about how this cyclic dependency is handled?

Reply:
Not particularly. I don't *love* the way this is handled (in terms of the two steps). If there was an easier way to construct the cycle (`Arc::new_cyclic...`) I wouldn't be opposed to it. But, it works, and it doesn't use excess trickery. It won't ever actually keep things around since the `new_cyclic` ensures that the cycles go through `Weak` references.

And the messiness only affects the `Pipeline` implementations, which we should have relatively few of (and it's far from the hardest part about writing the pipeline implementations).